### PR TITLE
Load and execute command

### PR DESCRIPTION
**Pull request by jziegel**
Adds a global command to find and load all files with the `*.js` extension

Issue https://github.com/creately/tester/issues/2

### CODE REVIEW

File: bin/tester.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env node
+
+const colors = require('colors');
+const test = require('../lib/test');
+
+const path = '/Users/jerome/projects/testfiles/';
```
Comment (mnmtanish) [inline bin/tester.js]: You can use `process.cwd` to get the working directory.

File: lib/test.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const globby = require('globby');
+const colors = require('colors');
+ 
+async function findFiles(directory) {
+    console.log('Searching for files in: '.yellow + directory);
```
Comment (mnmtanish) [inline lib/test.js]: Let's keep all console logs inside `bin/tester.js`.

File: lib/test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const globby = require('globby');
```
Comment (mnmtanish) [inline lib/test.js]: This repo is based on https://github.com/creately/ts-starter so we can use TypeScript inside `./src` directory. Let's move this file to `src` and use TypeScript.

File: bin/tester.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env node
+
+const colors = require('colors');
+const test = require('../lib/test');
```
Comment (mnmtanish) [inline bin/tester.js]: If we move the `lib/test` to `src` then this import will be `../dist/test`.

We can use JS for `bin/tester` because we'll be writing most of the code inside `src`.

File: lib/test.js (hunk 2 above)
Comment (jziegel) [reply lib/test.js]: Yep will do

File: bin/tester.js (hunk 1 above)
Comment (jziegel) [reply bin/tester.js]: Thanks 👍

File: lib/test.js (hunk 3 above)
Comment (jziegel) [reply lib/test.js]: Done

File: lib/test.js (hunk 2 above)
Comment (jziegel) [reply lib/test.js]: Done

File: bin/tester.js (hunk 4 above)
Comment (jziegel) [reply bin/tester.js]: Now the global `tester` command is linked to a script in the `dist` folder, I'm not sure if that's a good idea or not. But I like keeping the source in TypeScript.

File: src/__tests.ts [hunk 5]
```diff
@@ -1,3 +1,3 @@
-declare var require: any;
-const context = require.context('./', true, /\.spec\.ts$/);
+declare var noderequire: any;
+const context = noderequire.context('./', true, /\.spec\.ts$/);
```
Comment (mnmtanish) [inline src/__tests.ts]: Why change this? are unit tests working after this change?

File: src/tester.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env node
+
+import colors = require('colors');
```
Comment (mnmtanish) [inline src/tester.ts]: `import` or `require` ? Let's use `import`

File: src/tester.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    files.forEach((file: string) => {
+      console.log('Loading file: '.green + file);
+      if (file) require(path + file);
+    });
+  } else {
+    console.warn('No files found'.yellow);
+  }
+})().catch(err => console.error('Error: '.red, err));
+
+export {};
```
Comment (mnmtanish) [inline src/tester.ts]: This line is not needed.

File: bin/tester.js (hunk 4 above)
Comment (mnmtanish) [reply bin/tester.js]: How about putting commands in `src/bin` ?

File: src/__tests.ts (hunk 5 above)
Comment (jziegel) [reply src/__tests.ts]: It was causing an issue on `build` because of a duplicate declaration on `node_modules/@types/node/globals.d.ts:222:13`: `declare var require: NodeRequire;`. Unit tests are working though.

File: src/tester.ts (hunk 6 above)
Comment (jziegel) [reply src/tester.ts]: Done

File: src/tester.ts (hunk 7 above)
Comment (jziegel) [reply src/tester.ts]: Done

File: bin/tester.js (hunk 4 above)
Comment (jziegel) [reply bin/tester.js]: Sorry my bad, I forgot the dist folder is packaged by npm, all good

File: .npmignore [hunk 8]
```diff
@@ -1,8 +0,0 @@
-src
```
Comment (mnmtanish) [inline .npmignore]: Why is this file deleted? we can ignore these files when we publish it to npm to reduce the bundle size.

File: src/main.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import * as globby from 'globby';
+
+async function findFiles(path: string): Promise<string[]> {
```
Comment (mnmtanish) [inline src/main.ts]: Please use ES module syntax instead of `module.exports`

```ts
async function findFiles(path: string): Promise<string[]> {
  // ...
}
```

Comment (mnmtanish) [reply src/main.ts]: Add a comment for this function

File: src/tester.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node
+
+import 'colors';
+const test = require('./main');
```
Comment (mnmtanish) [inline src/tester.ts]: Use ES module syntax

```ts
import { findFiles } from `./main`
```

File: src/tester.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+(async () => {
+  console.log('Searching for files in: '.yellow + path);
+  const files: string[] = await test.findFiles(path);
+
+  if (files && files.length > 0) {
+    console.log('Found files: '.green + files);
+
+    files.forEach((file: string) => {
+      console.log('Loading file: '.green + file);
+      require(path + file);
```
Comment (mnmtanish) [inline src/tester.ts]: NOTE: this is a special case where using `require` makes more sense.

Comment (jziegel) [reply src/tester.ts]: Noted, thanks 👍

File: .npmignore (hunk 8 above)
Comment (jziegel) [reply .npmignore]: I've whitelisted the `dist/**/*` directory using the `files` key on `package.json`, that way it's easier to manage the package rather than blacklisting any new files we add. I did a dry run of `npm publish` and it seems to be packaging everything we require as expected.

https://blog.npmjs.org/post/165769683050/publishing-what-you-mean-to-publish

File: src/main.ts (hunk 9 above)
Comment (jziegel) [reply src/main.ts]: Done

File: src/tester.ts (hunk 10 above)
Comment (jziegel) [reply src/tester.ts]: Done

File: src/tester.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node
+
+import 'colors';
+import { findFiles } from './main';
+import yargs = require('yargs');
```
Comment (mnmtanish) [inline src/tester.ts]: This is not correct. It should be like

```ts
import yargs from 'yargs';
```

or 

```ts
import * as yargs from 'yargs';
```

File: src/main.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Finds all files in the given path that match the given extension.
+ * Looks through directories and sub directories. Default extension is '.test.js'.
+ * @param path the path to search in
+ * @param extensions a string array of extensions to look for
+ * @returns a promise with an array of paths to found files
+ */
+
+export async function findFiles(path: string, extensions: string[] = ['.test.js']): Promise<string[]> {
```
Comment (mnmtanish) [inline src/main.ts]: Please call `npm run lint` it will format the code using **prettier**.
https://prettier.io

Comment (jziegel) [reply src/main.ts]: Done, thanks

File: src/tester.ts (hunk 12 above)
Comment (jziegel) [reply src/tester.ts]: Done, thanks. The second is the syntax I should have used.
